### DISCUSSION TITLE
How to sum rows and columns of binary image?

### DISCUSSION THREAD

**Discussion by citroboy** [libvips/libvips#4214] [Help wanted]
I have a small black and white image.
I want to find the coords of the white area in the middle so I can crop it. I thought I would find the edges by adding up the values of the pixels in each row and column, but I see no obvious way to do that. I'm using pyvips.

**Comment by jcupitt** [comment]
Hello @citroboy, you need `project`:

https://www.libvips.org/API/current/libvips-arithmetic.html#vips-project

It scans the image once and makes two histograms, one of the row sums and one of the column sums.

```
Python 3.12.3 (main, Sep 11 2024, 14:17:37) [GCC 13.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyvips
>>> x = pyvips.Image.new_from_file("/home/john/pics/k2.jpg")
>>> x
<pyvips.Image 1450x2048 uchar, 3 bands, srgb>
>>> x.project()
[<pyvips.Image 1450x1 uint, 3 bands, histogram>, <pyvips.Image 1x2048 uint, 3 bands, histogram>]
>>> 
```

**Comment by citroboy** [comment]
Thanks for the quick response.

project seems to be oring rather than adding. The output images are black and white, so if there is 1 or 100 white pixels in a row I get white. Since there is some noise in the background that throws me off. I'm looking for a count of white pixels.

**Reply by jcupitt** [reply]
Project computes the sum of pixel values in that row or column. Do you want the number of non-zero pixels?

**Reply by jcupitt** [reply]
You could count non-zero pixels in rows and columns with something like:

```python
cols, rows = (x[1] > 128).ifthenelse(1, 0).project()
```

That's taking the green band, finding pixels greater than 128, and setting them to 1, with other pixels as 0. Then you project that to get the number of non-zero pixels in each row and column.

**Reply by citroboy** [reply]
Yes. project is outputting only 0 or 1 (like the input). That won't work for me.

**Reply by jcupitt** [reply]
It should work. I see:

```
$ python3
Python 3.12.3 (main, Sep 11 2024, 14:17:37) [GCC 13.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyvips
>>> x = pyvips.Image.new_from_file("/home/john/pics/k2.jpg")
>>> cols, rows = (x[1] > 128).ifthenelse(1, 0).project()
>>> cols.max()
1304.0
>>> 
```

So one column has 1304 green pixels which are greater than 128.

**Reply by jcupitt** [reply]
libvips uses 0 for false and 255 for true, so you could also write:

```
>>> cols, rows = ((x[1] > 128) >> 7).project()
>>> cols.max()
1304.0
>>> 
```

Which is a bit quicker, though maybe harder to read heh.

**Reply by citroboy** [reply]
`x.ifthenelse(1, 0).project()`

seems to have given some useful output. Thanks for the help, and the excellent mindbending library.
